manta-pay/src/crypto/hash/poseidon/parameter_generation/mds.rs

```diff
@@ -0,0 +1,412 @@
+            None => false,
+        }
+    }
+
+    /// Generates sparse matrix from m_double_prime matrix
+    pub fn new(m_double_prime: SquareMatrix<F>) -> Self
+    where
+        F: Clone,
+    {
+        assert!(Self::is_sparse(&m_double_prime));
+        let m_double_prime = Matrix::from(m_double_prime);
+        let w_hat = m_double_prime.iter_rows().map(|r| r[0].clone()).collect();
+        let v_rest = m_double_prime[0][1..].to_vec();
+        Self { w_hat, v_rest }
+    }
+
+    /// Size of the sparse matrix
+    pub fn size(&self) -> usize {
+        self.w_hat.len()
+    }
+
+    /// Generates dense-matrix representation from sparse matrix representation
+    pub fn to_matrix(&self) -> Matrix<F>
+    where
+        F: Clone,
+    {
+        let mut m = Matrix::identity(self.size());
+        for (j, elt) in self.w_hat.iter().enumerate() {
+            m[j][0] = elt.clone();
+        }
+        for (i, elt) in self.v_rest.iter().enumerate() {
+            m[0][i + 1] = elt.clone();
+        }
+        m
+    }
+}
+
+/// Factorizes into sparse matrices.
+pub fn factor_to_sparse_matrixes<F>(
+    base_matrix: SquareMatrix<F>,
+    n: usize,
+) -> (SquareMatrix<F>, Vec<SparseMatrix<F>>)
+where
+    F: Clone + Field + FieldGeneration + PartialEq,
+{
+    let (pre_sparse, mut sparse_matrices) =
+        (0..n).fold((base_matrix.clone(), Vec::new()), |(curr, mut acc), _| {
+            let derived = MdsMatrices::derive_mds_matrices(curr);
+            acc.push(derived.m_double_prime);
+            let new = base_matrix.matmul(&derived.m_prime).unwrap();
+            (new, acc)
+        });
+    sparse_matrices.reverse();
+    let sparse_matrices = sparse_matrices.into_iter().map(SparseMatrix::new).collect();
+    (pre_sparse, sparse_matrices)
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use ark_bls12_381::Fr;
+    use ark_ff::field_new;
+    use ark_std::{test_rng, UniformRand};
+
+    use crate::crypto::{
+        constraint::arkworks::Fp, hash::poseidon::parameter_generation::matrix::Matrix,
+    };
+
+    #[test]
+    fn mds_matrices_creation_is_correct() {
+        for i in 2..5 {
+            mds_matrices_creation_aux(i);
+        }
+    }
+
+    fn mds_matrices_creation_aux(width: usize) {
+        let MdsMatrices {
+            m,
+            m_inv,
+            m_hat,
+            m_hat_inv: _,
+            m_prime,
+            m_double_prime,
+        } = MdsMatrices::<Fp<Fr>>::new(width).unwrap();
+
+        for i in 0..m_hat.num_rows() {
+            for j in 0..m_hat.num_columns() {
+                assert_eq!(m[i + 1][j + 1], m_hat[i][j], "MDS minor has wrong value.");
+            }
+        }
+
+        // M^-1 x M = I
+        assert!(m_inv.matmul(&m).unwrap().is_identity());
+
+        // M' x M'' = M
+        assert_eq!(m, m_prime.matmul(&m_double_prime).unwrap());
+    }
+
+    #[test]
+    fn swapping_is_correct() {
+        swapping_aux(3)
+    }
+
+    fn swapping_aux(width: usize) {

```

Comment:
add doc string, change name, remove whitespace, and remove type annotations where possible